### ISSUE TITLE
Workers timeouts - gunicorns hangs up at recvfrom 

### ISSUE THREAD

**Issue by davies** [benoitc/gunicorn#517]
We have a box set up with Nginx 1.2.1 + Gunicorn 0.7.2. 

Under higher traffic (about 100+ qps), the workers will hangs up at recvfrom(), after 30 seconds, they will be killed.

```
17:43:47 sendto(64, "get cat.n_recipes:40076 \r\n", 26, MSG_DONTWAIT|MSG_NOSIGNAL, NULL, 0) = 26 <0.000049>
17:43:47 recvfrom(64, "VALUE cat.n_recipes:40076 2 5\r\n1"..., 8196, MSG_DONTWAIT, NULL, NULL) = 43 <0.000071>
17:43:47 sendto(63, "HTTP/1.0 200 OK\r\nServer: gunicor"..., 247, 0, NULL, 0) = 247 <0.000105>
17:43:47 sendto(63, "{\"xiachufang\": {\"@status\": \"ok\","..., 182, 0, NULL, 0) = 182 <0.000070>
17:43:47 stat("/etc/localtime", {st_mode=S_IFREG|0644, st_size=405, ...}) = 0 <0.000025>
17:43:47 close(63)                      = 0 <0.000091>
17:43:47 getppid()                      = 25286 <0.000033>
17:43:47 fchmod(52, 0)                  = 0 <0.000031>
17:43:47 select(63, [6], [], [61 62], {15, 0}) = 1 (in [6], left {14, 999994}) <0.000025>
17:43:47 fchmod(52, 01)                 = 0 <0.000025>
17:43:47 accept(6, {sa_family=AF_INET, sin_port=htons(35603), sin_addr=inet_addr("192.168.2.61")}, [16]) = 63 <0.000026>
17:43:47 fcntl(63, F_GETFL)             = 0x2 (flags O_RDWR) <0.000060>
17:43:47 fcntl(63, F_SETFL, O_RDWR)     = 0 <0.000016>
17:43:47 fcntl(63, F_GETFD)             = 0 <0.000016>
17:43:47 fcntl(63, F_SETFD, FD_CLOEXEC) = 0 <0.000068>
17:43:47 recvfrom(63,  <unfinished ...>
17:44:17 +++ killed by SIGKILL +++
```

The connection (fd = 63) comes from Nginx, in the error_log of nginx, we can see:

```
2013/04/24 17:44:17 [error] 20451#0: *13789663 recv() failed (104: Connection reset by peer) while reading response header from upstream, client: 118.145.8.14, server: api.xiachufang.com, request: "GET /?alt=json&method=recipe.getcomments&rid=164501 HTTP/1.1", upstream: "http://192.168.2.61:33302/api/?alt=json&method=recipe.getcomments&rid=164501", host: "api.xiachufang.com"
```

At the same time, It seems that Nginx and Gunicorn were both waiting for each other, and were blocked by each other.

Is this bug of gunicorn of Nginx?

**Comment by zoharby** [comment]
We've experienced the exact same issue a few days ago when our system was under higher than normal load after recovering from a  few hours of downtime..
All of our traffic is https. 

Running Nginx 1.6 and gunicorn 19.1.1

After a few restarts to the system, it suddenly started working properly, though we're not really sure what changed.

This seems to be the exact same case described here:
http://ppp.cylab.cmu.edu/wordpress/?p=1203

Bottom line, seems like this bug is still very relevant.

**Comment by benoitc** [comment]
what was the load?

**Comment by zoharby** [comment]
Around 10,000 RPM, which were divided by Google network load balancing between 2 VMS, each has 2 CPUs and 1.8 GB RAM (so around 5,000 RPM per machine).

**Comment by benoitc** [comment]
I don't think the load is the issue then, gunicorn is used with website having a larger traffic.

What is the version of python? Do you have a recent trace? Did you tested it with the latest gunicorn (maybe with the thread workers?)

**Comment by odedfos** [comment]
I'm answering on behalf of @zoharby,
python version is 2.7.3
If it's of any importance, the server is django nonrel v1.5.11.
We haven't tested yet with latest version nor with number of threads greater than 1 (with 5 workers)
